examples/tests/v3/test_v3_http_provider.py

```diff
@@ -0,0 +1,118 @@
+"""
+Test the FastAPI provider with Pact.
+
+This module tests the FastAPI provider defined in `src/fastapi.py` against the
+mock consumer. The mock consumer is set up by Pact and will replay the requests
+defined by the consumers. Pact will then validate that the provider responds
+with the expected responses.
+
+The provider will be expected to be in a given state in order to respond to
+certain requests. For example, when fetching a user's information, the provider
+will need to have a user with the given ID in the database. In order to avoid
+side effects, the provider's database calls are mocked out using functionalities
+from `unittest.mock`.
+
+In order to set the provider into the correct state, this test module defines an
+additional endpoint on the provider, in this case `/_pact/provider_states`.
+Calls to this endpoint mock the relevant database calls to set the provider into
+the correct state.
+"""
+
+from __future__ import annotations
+
+import time
+from multiprocessing import Process
+from typing import Dict, Union
+from unittest.mock import MagicMock
+
+import uvicorn
+from yarl import URL
+
+from examples.src.fastapi import app
+from pact.v3 import Verifier
+
+PROVIDER_URL = URL("http://localhost:8000")
+
+
+@app.post("/_pact/callback")
+async def mock_pact_provider_states(
+    action: str,  # noqa: ARG001
+    state: str,
+) -> Dict[str, Union[str, None]]:
+    """
+    Define the provider state.
+
+    For Pact to be able to correctly tests compliance with the contract, the
+    internal state of the provider needs to be set up correctly. Naively, this
+    would be achieved by setting up the database with the correct data for the
+    test, but this can be slow and error-prone. Instead this is best achieved by
+    mocking the relevant calls to the database so as to avoid any side effects.
+
+    For Pact to be able to correctly get the provider into the correct state,
+    this function is used to define an additional endpoint on the provider. This
+    endpoint is called by Pact before each test to ensure that the provider is
+    in the correct state.
+    """
+    mapping = {
+        "user doesn't exists": mock_user_doesnt_exist,
+        "user exists": mock_user_exists,
+    }
+    return {"result": mapping[state]()}
```

Comment:
I actually think we should handle the teardown in this situation.

As part of the teardown, I would want to check that the mocked database had the expected calls made to it.

For example, if we are expecting the request to perform a `FAKE_DB.get(1)`, then as part of the teardown, we should probably verify that the `get` method was indeed called, with the expected argument.

This would require the addition of two new functions: `verify_user_doesnt_exist_mock` and `verify_user_exists_mock`, and we would need to create a separate mapping based on whether the setup or teardown callback action is being sent.